strongbox-cron-tasks/src/main/java/org.carlspring.strongbox/cron/api/jobs/RebuildMetadataCronJob.java

```diff
@@ -0,0 +1,171 @@
+        try
+        {
+            String storageId = config.getProperty("storageId");
+            String repositoryId = config.getProperty("repositoryId");
+            String basePath = config.getProperty("basePath");
+
+            if (storageId == null)
+            {
+                List<String> storagesId = getStoragesId();
+                for (String storage : storagesId)
+                {
+                    Path storagePath = Paths.get(getStoragesBasePath(), storage);
+                    List<String> repositoriesId = getRepositoriesId(storagePath);
+                    rebuildRepositories(repositoriesId, storage, basePath);
+                }
+            }
+            else if (repositoryId == null)
+            {
+                Path storagePath = Paths.get(getStoragesBasePath(), storageId);
+                List<String> repositoriesId = getRepositoriesId(storagePath);
+                rebuildRepositories(repositoriesId, storageId, basePath);
+            }
+            else
+            {
+                artifactMetadataService.rebuildMetadata(storageId, repositoryId, basePath);
+            }
+        }
+        catch (IOException | XmlPullParserException | NoSuchAlgorithmException e)
+        {
+            e.printStackTrace();
+            logger.error("IOException: ", e);
+        }
+        manager.addExecutedJob(config.getName(), true);
+    }
+
+    /**
+     * To rebuild artifact's metadata in repositories
+     *
+     * @param repositories list of repositories
+     * @param storageId    path of storage
+     * @param basePath     basePath of artifact
+     * @throws NoSuchAlgorithmException
+     * @throws XmlPullParserException
+     * @throws IOException
+     */
+    private void rebuildRepositories(List<String> repositories,
+                                     String storageId,
+                                     String basePath)
+            throws NoSuchAlgorithmException, XmlPullParserException, IOException
+    {
+        for (String repository : repositories)
+        {
+            artifactMetadataService.rebuildMetadata(storageId, repository, basePath);
+        }
+    }
+
+    /**
+     * To get list of repositoryId in storage
+     *
+     * @param storage
+     * @return list of repositoryId
+     */
+    private List<String> getRepositoriesId(Path storage)
+    {
+        File file = storage.toFile();
+        return getListDirectories(file);
+    }
+
+    /**
+     * To get list of storageId
+     *
+     * @return list of storageId
+     */
+    private List<String> getStoragesId()
+    {
+        File file = Paths.get(storagesBasePath).toFile();
+        return getListDirectories(file);
+    }
+
+    /**
+     * To get list of directories
+     *
+     * @param baseFile Object File
+     * @return list of directories
+     */
+    private List<String> getListDirectories(File baseFile)
```

Comment:
I don't know, if I'm following your code properly, but I think a much more efficient and much more precise way of locating all the directories which contain artifacts would be using the [ArtifactDirectoryLocator](https://github.com/strongbox/strongbox/blob/master/strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/artifact/locator/ArtifactDirectoryLocator.java), for examples, you can check [ArtifactDirectoryLocatorTest](https://github.com/strongbox/strongbox/blob/master/strongbox-storage/strongbox-storage-api/src/test/java/org/carlspring/strongbox/storage/repository/artifact/locator/ArtifactDirectoryLocatorTest.java). This is all already being used by the [ArtifactMetadataServiceImpl.java](https://github.com/strongbox/strongbox/blob/master/strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactMetadataServiceImpl.java).